compiler/rustc_ty_utils/src/abi.rs

```diff
@@ -276,6 +277,25 @@ fn adjust_for_rust_scalar<'tcx>(
             }
         }
     }
+
+    // If this is the argument to `drop_in_place`, the contents of which we fully control as the
+    // compiler, then we may be able to mark that argument `noalias`. Currently, we're conservative
+    // and do so only if `drop_in_place` results in a direct call to the programmer's `drop` method.
+    // The `drop` method requires `&mut self`, so we're effectively just propagating the `noalias`
+    // guarantee from `drop` upward to `drop_in_place` in this case.
```

Comment:
Why doesn't this fall out of the deduced parameter attribute analysis done on MIR? Can we make it happen there instead of special casing Abi computation further?

Reply:
`deduced_param_attrs` only runs on user written functions, not on shims generated by the compiler. In addition it takes a `DefId`, but for `drop_in_place` the generated shim is unique for each generic argument.

Reply:
In addition to what @bjorn3 said, `deduced_param_attrs` is slower than doing the attribute assignment here, because it adds a query and has to write the results into the crate metadata. By doing the attribute assignment here, we can avoid that overhead for attributes that can be determined solely from the function identity and signature. This is the case for the pointer on `drop_in_place`—we know what the semantics of that function are immediately without having to inspect any MIR, and we can inspect the type of the parameter being dropped.